{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e439f8e6",
   "metadata": {},
   "source": [
    "# Landsat annual composites\n",
    "\n",
    "1. fetch the Landsat scenes overlapping our study region\n",
    "2. aggregate the Landsat scenes into a single multi-temporal composite using the median, our composite will contain the median reflectance obtained from all Landsat scenes fetched in certain year."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4dd12ebc-7150-4ace-bd3d-23a3ed8da625",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# we have to install the development version for the time being\n",
    "#!pip uninstall eodal -y\n",
    "#!pip install git+https://github.com/lukasValentin/eodal@landsat-dev\n",
    "#!pip install --upgrade git+https://github.com/EOA-team/eodal\n",
    "#!pip install --upgrade planetary-computer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1ed0dc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the year/years for the download\n",
    "\n",
    "#multiyears\n",
    "#years = list(range(2004, 2023))\n",
    "#excluded_years = [1988, 1992, 1993, 1996, 1997]\n",
    "#filtered_years = [year for year in years if year not in excluded_years]\n",
    "\n",
    "#single year\n",
    "filtered_years = [2022]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0b6a91cc-7b8c-402c-9df9-7cecc3e6bef7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import libraries\n",
    "from datetime import datetime\n",
    "from pathlib import Path\n",
    "from shapely.geometry import box\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "from eodal.config import get_settings\n",
    "from eodal.core.sensors import Landsat\n",
    "from eodal.mapper.feature import Feature\n",
    "from eodal.mapper.filter import Filter\n",
    "from eodal.mapper.mapper import Mapper, MapperConfigs\n",
    "from eodal.core.band import GeoInfo, Band\n",
    "from eodal.core.raster import RasterCollection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e87cb01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to mask clouds and shadows\n",
    "def preprocess_landsat_scene(\n",
    "        ds: Landsat\n",
    ") -> Landsat:\n",
    "    \"\"\"\n",
    "    Mask clouds and cloud shadows in a Landsat scene based\n",
    "    on the 'qa_pixel' band.\n",
    "\n",
    "    NOTE:\n",
    "        Depending on your needs, the pre-processing function can be\n",
    "        fully customized using the full power of EOdal and its\n",
    "        interfacing libraries!\n",
    "\n",
    "    :param ds:\n",
    "        Landsat scene before cloud mask applied.\n",
    "    :return:\n",
    "        Landsat scene with clouds and cloud shadows masked.\n",
    "    \"\"\"\n",
    "    ds.mask_clouds_and_shadows(inplace=True)\n",
    "    return ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "34ef1f09",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-14 10:33:23,186 eodal        INFO     Starting extraction of landsat scenes\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022 Number of Landsat scenes found: 122\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-14 10:33:32,977 eodal        WARNING  Scene with ID LE07_L2SP_173060_20220103_20220129_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:33:39,907 eodal        WARNING  Scene with ID LC09_L2SP_173059_20220103_20220122_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:33:44,342 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220103_20220122_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:33:55,747 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220111_20220122_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:34:13,779 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220119_20220122_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:34:31,830 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220127_20220204_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:34:48,749 eodal        WARNING  Scene with ID LE07_L2SP_173060_20220204_20220302_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:34:55,213 eodal        WARNING  Scene with ID LC09_L2SP_173059_20220204_20220206_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:34:59,642 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220204_20220206_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:35:17,466 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220212_20220222_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:35:30,424 eodal        WARNING  Scene with ID LC09_L2SP_172060_20220213_20220215_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:35:40,695 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220220_20220223_02_T2 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:36:04,511 eodal        WARNING  Scene with ID LE07_L2SP_173060_20220308_20220403_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:36:22,464 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220316_20220322_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:36:38,139 eodal        WARNING  Scene with ID LE07_L2SP_173060_20220324_20220419_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:36:44,566 eodal        WARNING  Scene with ID LC09_L2SP_173059_20220324_20220326_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:36:48,667 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220324_20220326_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:36:59,406 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220401_20220406_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:37:12,885 eodal        WARNING  Scene with ID LC09_L2SP_172060_20220402_20220404_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:37:24,062 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220409_20220411_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:37:44,988 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220417_20220427_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:37:55,755 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220425_20220427_02_T2 already added to SceneCollection - continue\n",
      "2023-07-14 10:38:14,798 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220503_20220511_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band qa_aerosol not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:38:28,723 eodal        WARNING  Scene with ID LE07_L2SP_173060_20220510_20220901_02_T1 already added to SceneCollection - continue\n",
      "c:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:221: UserWarning: Band cloud_qa not found - remove it from selection\n",
      "  warnings.warn(\n",
      "2023-07-14 10:38:39,667 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220511_20220513_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:38:51,249 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220519_20220525_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:39:02,515 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220527_20220529_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:39:20,859 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220604_20220610_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:39:39,373 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220612_20220614_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:39:58,257 eodal        WARNING  Scene with ID LC08_L2SP_173060_20220620_20220701_02_T1 already added to SceneCollection - continue\n",
      "2023-07-14 10:40:19,055 eodal        WARNING  Scene with ID LC09_L2SP_173060_20220628_20220630_02_T1 already added to SceneCollection - continue\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Could not load scene:  Could not add band red from LC08_L2SP_173059_20220706_20220722_02_T1: Cannot initialize new Band instance: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mCPLE_OpenFailedError\u001b[0m                      Traceback (most recent call last)",
      "File \u001b[1;32mrasterio\\_base.pyx:310\u001b[0m, in \u001b[0;36mrasterio._base.DatasetBase.__init__\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mrasterio\\_base.pyx:221\u001b[0m, in \u001b[0;36mrasterio._base.open_dataset\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mrasterio\\_err.pyx:221\u001b[0m, in \u001b[0;36mrasterio._err.exc_wrap_pointer\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mCPLE_OpenFailedError\u001b[0m: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mRasterioIOError\u001b[0m                           Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\raster.py:1002\u001b[0m, in \u001b[0;36mRasterCollection.add_band\u001b[1;34m(self, band_constructor, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1001\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m-> 1002\u001b[0m         band \u001b[39m=\u001b[39m band_constructor\u001b[39m.\u001b[39m\u001b[39m__call__\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1003\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\band.py:762\u001b[0m, in \u001b[0;36mBand.from_rasterio\u001b[1;34m(cls, fpath_raster, band_idx, band_name_src, band_name_dst, vector_features, full_bounding_box_only, epsg_code, **kwargs)\u001b[0m\n\u001b[0;32m    761\u001b[0m masking \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m--> 762\u001b[0m gdf_aoi \u001b[39m=\u001b[39m check_aoi_geoms(\n\u001b[0;32m    763\u001b[0m     in_dataset\u001b[39m=\u001b[39;49mvector_features,\n\u001b[0;32m    764\u001b[0m     fname_raster\u001b[39m=\u001b[39;49m_fpath_raster,\n\u001b[0;32m    765\u001b[0m     full_bounding_box_only\u001b[39m=\u001b[39;49mfull_bounding_box_only,\n\u001b[0;32m    766\u001b[0m     raster_crs\u001b[39m=\u001b[39;49mepsg_code\n\u001b[0;32m    767\u001b[0m )\n\u001b[0;32m    768\u001b[0m \u001b[39m# check for third dimension (has_z) and flatten it to 2d\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\utils\\reprojection.py:135\u001b[0m, in \u001b[0;36mcheck_aoi_geoms\u001b[1;34m(in_dataset, full_bounding_box_only, fname_raster, raster_crs)\u001b[0m\n\u001b[0;32m    134\u001b[0m \u001b[39mif\u001b[39;00m fname_raster \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m--> 135\u001b[0m     sat_crs \u001b[39m=\u001b[39m rio\u001b[39m.\u001b[39;49mopen(fname_raster)\u001b[39m.\u001b[39mcrs\n\u001b[0;32m    136\u001b[0m \u001b[39m# if the raster has no inherent CRS use the user-defined one\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\rasterio\\env.py:451\u001b[0m, in \u001b[0;36mensure_env_with_credentials.<locals>.wrapper\u001b[1;34m(*args, **kwds)\u001b[0m\n\u001b[0;32m    450\u001b[0m \u001b[39mwith\u001b[39;00m env_ctor(session\u001b[39m=\u001b[39msession):\n\u001b[1;32m--> 451\u001b[0m     \u001b[39mreturn\u001b[39;00m f(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwds)\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\rasterio\\__init__.py:320\u001b[0m, in \u001b[0;36mopen\u001b[1;34m(fp, mode, driver, width, height, count, crs, transform, dtype, nodata, sharing, **kwargs)\u001b[0m\n\u001b[0;32m    319\u001b[0m \u001b[39mif\u001b[39;00m mode \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m--> 320\u001b[0m     dataset \u001b[39m=\u001b[39m DatasetReader(path, driver\u001b[39m=\u001b[39mdriver, sharing\u001b[39m=\u001b[39msharing, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m    321\u001b[0m \u001b[39melif\u001b[39;00m mode \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mr+\u001b[39m\u001b[39m\"\u001b[39m:\n",
      "File \u001b[1;32mrasterio\\_base.pyx:312\u001b[0m, in \u001b[0;36mrasterio._base.DatasetBase.__init__\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mRasterioIOError\u001b[0m: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:417\u001b[0m, in \u001b[0;36mLandsat.from_usgs\u001b[1;34m(cls, in_dir, band_selection, read_qa, read_atcor, apply_scaling, **kwargs)\u001b[0m\n\u001b[0;32m    416\u001b[0m         kwargs\u001b[39m.\u001b[39mupdate({\u001b[39m'\u001b[39m\u001b[39mscale\u001b[39m\u001b[39m'\u001b[39m: gain, \u001b[39m'\u001b[39m\u001b[39moffset\u001b[39m\u001b[39m'\u001b[39m: offset})\n\u001b[1;32m--> 417\u001b[0m     landsat\u001b[39m.\u001b[39madd_band(\n\u001b[0;32m    418\u001b[0m         Band\u001b[39m.\u001b[39mfrom_rasterio,\n\u001b[0;32m    419\u001b[0m         fpath_raster\u001b[39m=\u001b[39mband_fpath,\n\u001b[0;32m    420\u001b[0m         band_idx\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m,\n\u001b[0;32m    421\u001b[0m         band_name_dst\u001b[39m=\u001b[39mband_name,\n\u001b[0;32m    422\u001b[0m         band_alias\u001b[39m=\u001b[39mband_alias,\n\u001b[0;32m    423\u001b[0m         \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[0;32m    424\u001b[0m     )\n\u001b[0;32m    425\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\raster.py:1004\u001b[0m, in \u001b[0;36mRasterCollection.add_band\u001b[1;34m(self, band_constructor, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1003\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m-> 1004\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCannot initialize new Band instance: \u001b[39m\u001b[39m{\u001b[39;00me\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m   1006\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "\u001b[1;31mValueError\u001b[0m: Cannot initialize new Band instance: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mException\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\mapper\\mapper.py:429\u001b[0m, in \u001b[0;36mMapper._process_scene\u001b[1;34m(self, item, scene_constructor, scene_constructor_kwargs, scene_modifier, scene_modifier_kwargs, reprojection_method)\u001b[0m\n\u001b[0;32m    427\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    428\u001b[0m     \u001b[39m# call scene constructor. The file-path (or URL) goes first\u001b[39;00m\n\u001b[1;32m--> 429\u001b[0m     scene \u001b[39m=\u001b[39m scene_constructor\u001b[39m.\u001b[39m\u001b[39m__call__\u001b[39m(\n\u001b[0;32m    430\u001b[0m         item\u001b[39m.\u001b[39mreal_path, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mscene_constructor_kwargs\n\u001b[0;32m    431\u001b[0m     )\n\u001b[0;32m    432\u001b[0m     scene\u001b[39m.\u001b[39mscene_properties\u001b[39m.\u001b[39msensing_time \u001b[39m=\u001b[39m item[\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtime_column]\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\core\\sensors\\landsat.py:426\u001b[0m, in \u001b[0;36mLandsat.from_usgs\u001b[1;34m(cls, in_dir, band_selection, read_qa, read_atcor, apply_scaling, **kwargs)\u001b[0m\n\u001b[0;32m    425\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m--> 426\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\n\u001b[0;32m    427\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCould not add band \u001b[39m\u001b[39m{\u001b[39;00mband_name\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m\n\u001b[0;32m    428\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mfrom \u001b[39m\u001b[39m{\u001b[39;00mscene_properties\u001b[39m.\u001b[39mproduct_uri\u001b[39m}\u001b[39;00m\u001b[39m: \u001b[39m\u001b[39m{\u001b[39;00me\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[0;32m    429\u001b[0m     )\n\u001b[0;32m    430\u001b[0m \u001b[39m# apply actual vector features if masking is required\u001b[39;00m\n",
      "\u001b[1;31mException\u001b[0m: Could not add band red from LC08_L2SP_173059_20220706_20220722_02_T1: Cannot initialize new Band instance: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[10], line 69\u001b[0m\n\u001b[0;32m     62\u001b[0m scene_kwargs \u001b[39m=\u001b[39m {\n\u001b[0;32m     63\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mscene_constructor\u001b[39m\u001b[39m'\u001b[39m: Landsat\u001b[39m.\u001b[39mfrom_usgs,\n\u001b[0;32m     64\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mscene_constructor_kwargs\u001b[39m\u001b[39m'\u001b[39m: {\u001b[39m'\u001b[39m\u001b[39mband_selection\u001b[39m\u001b[39m'\u001b[39m: [\u001b[39m'\u001b[39m\u001b[39mblue\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mgreen\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mred\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mnir08\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mswir16\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mswir22\u001b[39m\u001b[39m'\u001b[39m], \u001b[39m'\u001b[39m\u001b[39mread_qa\u001b[39m\u001b[39m'\u001b[39m: \u001b[39mTrue\u001b[39;00m} ,\n\u001b[0;32m     65\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mscene_modifier\u001b[39m\u001b[39m'\u001b[39m: preprocess_landsat_scene,\n\u001b[0;32m     66\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mscene_modifier_kwargs\u001b[39m\u001b[39m'\u001b[39m: {}}\n\u001b[0;32m     68\u001b[0m \u001b[39m# now we load the scenes\u001b[39;00m\n\u001b[1;32m---> 69\u001b[0m mapper\u001b[39m.\u001b[39;49mload_scenes(scene_kwargs\u001b[39m=\u001b[39;49mscene_kwargs)\n\u001b[0;32m     71\u001b[0m \u001b[39m# The mapper returns the single scenes. \u001b[39;00m\n\u001b[0;32m     72\u001b[0m \u001b[39m# As we told the EOdal to mask out clouds, a significant share of the pixels is masked out. \u001b[39;00m\n\u001b[0;32m     73\u001b[0m \u001b[39m# We will aggregate them in the next step.\u001b[39;00m\n\u001b[0;32m     74\u001b[0m f \u001b[39m=\u001b[39m mapper\u001b[39m.\u001b[39mdata\u001b[39m.\u001b[39mplot(band_selection\u001b[39m=\u001b[39m[\u001b[39m'\u001b[39m\u001b[39mred\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mgreen\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mblue\u001b[39m\u001b[39m'\u001b[39m], figsize \u001b[39m=\u001b[39m (\u001b[39m20\u001b[39m, \u001b[39m20\u001b[39m), max_scenes_in_row \u001b[39m=\u001b[39m \u001b[39m2\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\mapper\\mapper.py:810\u001b[0m, in \u001b[0;36mMapper.load_scenes\u001b[1;34m(self, scene_kwargs, pixel_kwargs)\u001b[0m\n\u001b[0;32m    808\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_load_pixels(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mpixel_kwargs)\n\u001b[0;32m    809\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 810\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_load_scenes_collection(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mscene_kwargs)\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\mapper\\mapper.py:617\u001b[0m, in \u001b[0;36mMapper._load_scenes_collection\u001b[1;34m(self, reprojection_method, scene_constructor, scene_constructor_kwargs, scene_modifier, scene_modifier_kwargs)\u001b[0m\n\u001b[0;32m    615\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m _metadata_unique\u001b[39m.\u001b[39mempty:\n\u001b[0;32m    616\u001b[0m     \u001b[39mfor\u001b[39;00m _, item \u001b[39min\u001b[39;00m _metadata_unique\u001b[39m.\u001b[39miterrows():\n\u001b[1;32m--> 617\u001b[0m         scene \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_process_scene(\n\u001b[0;32m    618\u001b[0m             item\u001b[39m=\u001b[39;49mitem,\n\u001b[0;32m    619\u001b[0m             scene_constructor\u001b[39m=\u001b[39;49mscene_constructor,\n\u001b[0;32m    620\u001b[0m             scene_constructor_kwargs\u001b[39m=\u001b[39;49mscene_constructor_kwargs,\n\u001b[0;32m    621\u001b[0m             scene_modifier\u001b[39m=\u001b[39;49mscene_modifier,\n\u001b[0;32m    622\u001b[0m             scene_modifier_kwargs\u001b[39m=\u001b[39;49mscene_modifier_kwargs,\n\u001b[0;32m    623\u001b[0m             reprojection_method\u001b[39m=\u001b[39;49mreprojection_method,\n\u001b[0;32m    624\u001b[0m         )\n\u001b[0;32m    625\u001b[0m         \u001b[39m# because ESA has some mess with the naming of their file names and\u001b[39;00m\n\u001b[0;32m    626\u001b[0m         \u001b[39m# metadata, there might be duplicated seems not detected by the mapper\u001b[39;00m\n\u001b[0;32m    627\u001b[0m         \u001b[39m# since the time stamps slightly differ (few seconds in some cases) but\u001b[39;00m\n\u001b[0;32m    628\u001b[0m         \u001b[39m# their IDs are the same\u001b[39;00m\n\u001b[0;32m    629\u001b[0m         \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\ftimcke\\AppData\\Local\\anaconda3\\lib\\site-packages\\eodal\\mapper\\mapper.py:434\u001b[0m, in \u001b[0;36mMapper._process_scene\u001b[1;34m(self, item, scene_constructor, scene_constructor_kwargs, scene_modifier, scene_modifier_kwargs, reprojection_method)\u001b[0m\n\u001b[0;32m    432\u001b[0m     scene\u001b[39m.\u001b[39mscene_properties\u001b[39m.\u001b[39msensing_time \u001b[39m=\u001b[39m item[\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtime_column]\n\u001b[0;32m    433\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m--> 434\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCould not load scene:  \u001b[39m\u001b[39m{\u001b[39;00me\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    436\u001b[0m \u001b[39m# apply scene modifier callable if available\u001b[39;00m\n\u001b[0;32m    437\u001b[0m \u001b[39mif\u001b[39;00m scene_modifier \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[1;31mValueError\u001b[0m: Could not load scene:  Could not add band red from LC08_L2SP_173059_20220706_20220722_02_T1: Cannot initialize new Band instance: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/173/059/LC08_L2SP_173059_20220706_20220722_02_T1/LC08_L2SP_173059_20220706_20220722_02_T1_SR_B4.TIF?st=2023-07-13T08%3A11%3A43Z&se=2023-07-14T08%3A56%3A43Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2023-07-14T05%3A41%3A05Z&ske=2023-07-21T05%3A41%3A05Z&sks=b&skv=2021-06-08&sig=y9Ihand0qAQD/uYH3OnXz4owYfjrW9VwFI3j3mf34rc%3D' not recognized as a supported file format."
     ]
    }
   ],
   "source": [
    "## Setting up the EOdal Mapper\n",
    "Settings = get_settings()\n",
    "\n",
    "# we use STAC, i.e., Microsoft Planetary Computer\n",
    "Settings.USE_STAC = True\n",
    "\n",
    "# user-inputs\n",
    "# -------------------------- Collection -------------------------------\n",
    "collection = 'landsat-c2-l2'\n",
    "\n",
    "# ---------------------- Spatial Feature  ------------------------------\n",
    "# can be also shp, gpkg, etc.\n",
    "\n",
    "# bbox = box(*[30.2825, 0.4019, 30.3714, 0.4643]) # 1. Area\n",
    "# bbox = box(*[30.2703, 0.4043, 30.4482, 0.5291]) # 2. Area\n",
    "# bbox = box(*[30.3195, 0.4803, 30.3417, 0.4959]) # 3. Area small for MeanShift trial\n",
    "bbox = box(*[30.2433, 0.3746, 30.5640, 0.6962]) # 4. Area bigger\n",
    "\n",
    "feature = Feature(\n",
    "    name='landsat-composite',\n",
    "    geometry=bbox,\n",
    "    epsg=4326,\n",
    "    attributes={})\n",
    "\n",
    "# ------------------------- Metadata Filters ---------------------------\n",
    "metadata_filters = [\n",
    "    #Filter('eo:cloud_cover', '<=', 80),\n",
    "    #Filter('landsat:wrs_path', '==', '173'),\n",
    "    #Filter('landsat:wrs_row', '==', '060'),\n",
    "    #Filter('instruments', '!=', 'etm+')\n",
    "]\n",
    "\n",
    "for year in filtered_years: \n",
    "    \n",
    "    # ------------------------- Time Range ---------------------------------\n",
    "    time_start = datetime(year, 1, 1)\n",
    "    time_end = datetime(year, 12, 31)\n",
    "\n",
    "    # set up the Mapper configuration\n",
    "    mapper_configs = MapperConfigs(\n",
    "        metadata_filters=metadata_filters,\n",
    "        collection=collection,\n",
    "        feature=feature,\n",
    "        time_start=time_start,\n",
    "        time_end=time_end)\n",
    "\n",
    "    # get a new mapper instance\n",
    "    mapper = Mapper(mapper_configs)\n",
    "\n",
    "    # fetch the metadata\n",
    "    # query the scenes available (no I/O of scenes, this only fetches metadata)\n",
    "    mapper.query_scenes()\n",
    "    print(str(year), f'Number of Landsat scenes found: {mapper.metadata.shape[0]}')\n",
    "    \n",
    "    if mapper.metadata.empty:\n",
    "        print(f'no scenes found in year: {year}')\n",
    "        continue\n",
    "\n",
    "    #We tell EOdal how to load the Landsat scenes using `Landsat.from_usgs`and pass on some kwargs, e.g., the selection of bands we want to read.in addition, we tell EOdal to mask out clouds and shadows and the fly while reading the data using the qa_pixel band (therefore, we set the`read_qa` flag to True.\n",
    "    \n",
    "    # define the bands to read in 'band_selection' and to preprocess the scenes\n",
    "    scene_kwargs = {\n",
    "        'scene_constructor': Landsat.from_usgs,\n",
    "        'scene_constructor_kwargs': {'band_selection': ['blue', 'green', 'red', 'nir08', 'swir16', 'swir22'], 'read_qa': True} ,\n",
    "        'scene_modifier': preprocess_landsat_scene,\n",
    "        'scene_modifier_kwargs': {}}\n",
    "\n",
    "    # now we load the scenes\n",
    "    mapper.load_scenes(scene_kwargs=scene_kwargs)\n",
    "\n",
    "    # The mapper returns the single scenes. \n",
    "    # As we told the EOdal to mask out clouds, a significant share of the pixels is masked out. \n",
    "    # We will aggregate them in the next step.\n",
    "    f = mapper.data.plot(band_selection=['red', 'green', 'blue'], figsize = (20, 20), max_scenes_in_row = 2)\n",
    "\n",
    "    #save as PNG for quick view\n",
    "    f.savefig(f'S:\\MSc_23_TimckeFinn\\data\\python_outputs\\landsat_scenes_{year}.png')\n",
    "    \n",
    "    plt.close(f)  # Close the figure (uses less memory?)\n",
    "\n",
    "    ## Generating the composite, We will now generate the composite. Here, we have to consider two things: 1. The data is stored as [numpy masked arrays](https://numpy.org/doc/stable/reference/maskedarray.generic.html) as we masked out clouds. 2. We therefore have to use masked arrays and the numpy ma functions that work on numpy masked arrays (i.e., the ignore masked values).\n",
    "\n",
    "    #First, we open masked arrays for storing the data:\n",
    "    # all scenes have the same shape, i.e., the same number of bands, rows and columns\n",
    "    shapes = [{timestamp: scene.get_values().shape} for timestamp, scene in mapper.data]\n",
    "\n",
    "    # open arrays for storing the data per band\n",
    "    shape = (len(mapper.data), list(shapes[0].values())[0][1], list(shapes[0].values())[0][2])\n",
    "    blue = np.ma.masked_array(data=np.ndarray(shape, dtype=float), mask=False)\n",
    "    red = np.zeros_like(blue)\n",
    "    green = np.zeros_like(blue)\n",
    "    nir08 = np.zeros_like(blue)\n",
    "    swir16 = np.zeros_like(blue)\n",
    "    swir22 = np.zeros_like(blue)\n",
    "\n",
    "    # Next, we loop over the scenes.\n",
    "    idx = 0\n",
    "    for _, scene in mapper.data:\n",
    "        blue[idx, :, :] = scene['blue'].values\n",
    "        red[idx, :, :] = scene['red'].values\n",
    "        green[idx, :, :] = scene['green'].values\n",
    "        nir08[idx, :, :] = scene['nir08'].values\n",
    "        swir16[idx, :, :] = scene['swir16'].values\n",
    "        swir22[idx, :, :] = scene['swir22'].values\n",
    "        idx += 1\n",
    "\n",
    "    # Finally, we aggregate the data using the median reflectance\n",
    "    # calculate the median reflectance per spectral band\n",
    "    blue_median = np.ma.median(blue, axis=0)\n",
    "    green_median = np.ma.median(green, axis=0)\n",
    "    red_median = np.ma.median(red, axis=0)\n",
    "    nir08_median = np.ma.median(nir08, axis=0)\n",
    "    swir16_median = np.ma.median(swir16, axis=0)\n",
    "    swir22_median = np.ma.median(swir22, axis=0)\n",
    "\n",
    "    # We store the results in a new RasterCollection save the median reflectance to a new RasterCollection\n",
    "    rc = RasterCollection()\n",
    "\n",
    "    bands = {'blue': blue_median, \n",
    "            'green': green_median, \n",
    "            'red': red_median, \n",
    "            'nir08' : nir08_median, \n",
    "            'swir16' : swir16_median,\n",
    "            'swir22' : swir22_median}\n",
    "\n",
    "    for band_name, band_value in bands.items():\n",
    "        rc.add_band(\n",
    "            band_constructor=Band,\n",
    "            values=band_value,\n",
    "            band_name=f'{band_name}_median',\n",
    "            geo_info=scene[band_name].geo_info)\n",
    "        \n",
    "    # # We plot the result\n",
    "    f, ax = plt.subplots(figsize = (20, 10))\n",
    "    f = rc.plot_multiple_bands(['red_median', 'green_median', 'blue_median'], ax=ax)\n",
    "\n",
    "    #save as PNG for quick view\n",
    "    f.savefig(f'S:\\MSc_23_TimckeFinn\\data\\python_outputs\\landsat_median_composite_{year}.png')\n",
    "\n",
    "    plt.close(f)  # Close the figure (uses less memory?)\n",
    "        \n",
    "    # save as GeoTiff for further analysis\n",
    "    rc.to_rasterio(f'S:\\MSc_23_TimckeFinn\\data\\EOdal\\landsat_median_composite_{year}.tif')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
